### PR TITLE
LibIMAP+Mail: A bunch of fixes so I can load and browse my inbox :^)

### PR DESCRIPTION
**Pull request by vkoskiv**
I decided to try out `Mail` with my actual inbox, and found a bunch of crashes. This PR resolves some of those crashes, and adds some new functionality.

Demo!
image

Obviously I'm not going to show my whole inbox, but loading HTML messages works fine, and looks pretty legit!
`Mail` happily loaded my entire inbox of about ~24k message headers (Subject, From, Date) in about 53 seconds, most of which was just spent receiving the `IMAP` response from the GMail server.

Some of the new features include:
* We now support RFC-2047 message header decoding, so we can see special characters and emojis in email titles now!
* We now fetch and display the `Date` header of emails, and display that in the message view.
* I reworked the IMAP parser to be (very) slightly more robust. It now crashes with purpose, instead of accidentally.
* I added useful debugging prints to `IMAP::Parser`, available under the `IMAP_PARSER_DEBUG` flag.
* We now wait for a full `IMAP` response to arrive, before passing it to our parser. Our parser cannot deal with partial input yet.
* The status bar is updated as `IMAP::Client` is working, to keep the user in the loop about what's happening.
* Some obvious, and some less obvious fixes to crashes.

I think this is more than enough commits for a single PR, so I wish to tackle the following in the near future in other PRs:

- [ ] Use `LibSQL` to build a local cache, instead of fetching giant message header lists every time the user selects an inbox :^)
- [ ] Add support for sending emails.
- [ ] Add a flag to `LibWeb` to disallow fetching remote resources in messages until the user confirms.
- [ ] Make the IMAP parser more robust, and fully spec-compliant.

I'd also like to explore the possibility of a `KeychainService` or similar, so we can remove this very scary line of code from `MailWidget.cpp`:

`auto password = Config::read_string("Mail"sv, "User"sv, "Password"sv, {});`

### CODE REVIEW

**Comment by ADKaster** [comment]
Looks like you've got some compile errors. Did you maybe forget to add some local changes/fixups?

https://github.com/SerenityOS/serenity/actions/runs/5719789293/job/15498411591?pr=20290#step:25:4463

**Comment by vkoskiv** [comment]
> Looks like you've got some compile errors. Did you maybe forget to add some local changes/fixups?

Sure did! :sweat_smile: I had limited the amount of email headers to fetch for testing, and that local variable leaked in as I was building commits. Fixed now.

**Comment by vkoskiv** [comment]
Note for those thinking of testing this with a non-test account: #20292 resolves an issue where opening a mailbox sets the `\Seen` flag in all messages in that box to `true`. If you're testing without those changes applied, your unread emails will be marked as read.

**Comment by vkoskiv** [comment]
> Note for those thinking of testing this with a non-test account: #20292 resolves an issue where opening a mailbox sets the `\Seen` flag in all messages in that box to `true`. If you're testing without those changes applied, your unread emails will be marked as read.

^ I just rebased on `master`, so that particular issue is now resolved. Still though, buyer beware!

**Comment by vkoskiv** [comment]
Rebased on `master` again, fixed conflicts.

File: Userland/Applications/Mail/MailWidget.cpp [hunk 1]
```diff
@@ -368,10 +368,17 @@ void MailWidget::selected_mailbox()
...
+            from_index = from_iterator_value.find("from:"sv);
+        DeprecatedString from;
+        if (from_index.has_value()) {
+            auto potential_from = from_iterator_value.substring(from_index.value());
+            auto from_parts = potential_from.split_limit(':', 2);
+            from = parse_and_unfold(from_parts.last());
+        }
+
+        if (from.is_empty())
+            from = "(Unknown sender)";
```
Comment (BertalanD) [inline Userland/Applications/Mail/MailWidget.cpp]: What do the relevant RFCs say about the case sensitivity of email headers? Should we do a case-insensitive match instead?

Comment (vkoskiv) [reply Userland/Applications/Mail/MailWidget.cpp]: [RFC-2076](https://www.rfc-editor.org/rfc/rfc2076#section-3.4) sums it up nicely, the actual header should be: `From:`, but in practice, my inbox of ~25k messages contained one single message where that header was `from:`, and no messages where the `Subject:` field deviated from specifications.
I don't mind doing a case-insensitive match, but that falls beyond the scope of this PR, I don't think we have a `DeprecatedString::find()` overload that takes a `CaseSensitivity` input.

**Comment by vkoskiv** [comment]
Rebased on `master` again just to keep it up to date.

**Review by ADKaster — APPROVED** [review]

File: Userland/Libraries/LibIMAP/Client.cpp [hunk 2]
```diff
@@ -53,6 +53,26 @@ ErrorOr<NonnullOwnPtr<Client>> Client::connect_plaintext(StringView host, u16 po
...
 
+bool Client::verify_response_is_complete()
+{
+    // FIXME: This is still more of a heuristic than a proper approach.
+    //        I would imagine this breaks if we happen to get an email that
+    //        contains this pattern we're looking for.
+    dbgln("Waiting for a complete IMAP response, buffer size is now {}", m_buffer.size());
+    Vector<StringView> statuses = { "OK"sv, "BAD"sv, "NO"sv };
+    auto slice_size = m_buffer.size() >= 100 ? 100 : m_buffer.size(); // Arbitrary slice size, should contain what we're looking for.
+    auto slice_data = MUST(m_buffer.slice(m_buffer.size() - slice_size, slice_size));
```
[the remark this answers is not in the document]
Comment (vkoskiv) [reply Userland/Libraries/LibIMAP/Client.cpp:64]: It's trying to figure out if we have the full `IMAP` server response from the server before passing it to the parser. `IMAP::parser()` may blow up if given partial/incomplete IMAP response data.
It just checks if we've received one of those three status replies, which appear at the end of (hopefully) every IMAP response we get from the server.
I am quite sure this isn't the proper heuristic for this, but it worked reliably in my testing.

**Pull request merged**